Allow control plane migration across seeds w/ different provider types (#10323)
gardener/gardener: docs/operations/control_plane_migration.md

## Prerequisites

The `Seed`s involved in the control plane migration must have backups enabled - their `.spec.backup` fields cannot be nil.

## ShootState

`ShootState` is an API resource which stores non-reconstructible state and data required to completely recreate a `Shoot`'s control plane on a new `Seed`.  The `ShootState` resource is created on `Shoot` creation in its `Project` namespace and the required state/data is persisted during `Shoot` creation or reconciliation.

## Shoot Control Plane Migration

> Triggering the migration is done by changing the `Shoot`'s `.spec.seedName` to a `Seed` that differs from the `.status.seedName`, we call this `Seed` a `"Destination Seed"`. This action can only be performed by an operator with the necessary RBAC. If the `Destination Seed` does not have a backup and restore configuration, the change to `spec.seedName` is rejected. Additionally, this Seed must not be set for deletion and must be healthy.

Reviewer: ```suggestion
Triggering the migration is done by changing the `Shoot`'s `.spec.seedName` to a `Seed` that differs from the `.status.seedName`, we call this `Seed` a `"Destination Seed"`. This action can only be performed by an operator (see [Triggering the Migration](#triggering-the-migration)). If the `Destination Seed` does not have a backup and restore configuration, the change to `spec.seedName` is rejected. Additionally, this Seed must not be set for deletion and must be healthy.
```

Nit.

## Triggering the Migration

For control plane migration, operators with the necessary RBAC can use the [`shoots/binding`](../concepts/scheduler.md#shootsbinding-subresource) subresource to change the `.spec.seedName`, with the following commands:

```bash
NAMESPACE=my-namespace
SHOOT_NAME=my-shoot
DEST_SEED_NAME=destination-seed

kubectl get --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME} | jq -c '.spec.seedName = "'${DEST_SEED_NAME}'"' | kubectl replace --raw /apis/core.gardener.cloud/v1beta1/namespaces/${NAMESPACE}/shoots/${SHOOT_NAME}/binding -f - | jq -r '.spec.seedName'
```

> > [!NOTE]

Reviewer: ```suggestion
> [!IMPORTANT]
```

Nit.

>
> The nodes of your `Shoot` cluster must have network connectivity to the `Shoot`'s `kube-apiserver` and the `vpn-seed-server` once they are migrated to the `Destination Seed`. Otherwise, the `Restore` operation will get stuck at the `Waiting until the Kubernetes API server can connect to the Shoot workers` step. However, if you do end up in this case and cannot allow network traffic from the nodes to the `Shoot`'s control plane, you can annotate the `Shoot` with the `shoot.gardener.cloud/skip-readiness` annotation so that the `Restore` operation finishes, and then use the [`shoots/binding`](../concepts/scheduler.md#shootsbinding-subresource) subresource to migrate the control plane back to the `Source Seed`.

## Copying ETCD Backups Manually During the `Restore` Operation

> Following is a workaround that can be used to copy etcd backups manually in situations where a `Shoot`'s control plane has been moved to a `Destination Seed` and the pods running in it lack network connectivity to the `Source Seed`s storage provider:

Reviewer: ```suggestion
Following is a workaround that can be used to copy etcd backups manually in situations where a `Shoot`'s control plane has been moved to a `Destination Seed` and the pods running in it lack network connectivity to the `Source Seed`'s storage provider:
```

1. Use the `etcdbrctl copy` command to copy the backups by following the instructions in the [`etcdbrctl copy` guide](https://github.com/gardener/etcd-backup-restore/blob/master/docs/deployment/getting_started.md#etcdbrctl-copy)
1. After you have successfully copied the etcd backups, wait for the `EtcdCopyBackupsTask` custom resource to be created in the `Shoot`'s control plane on the `Destination Seed`, if it does not already exist. Afterwards, mark it as successful by patching it using the following command:
    ```
    SHOOT_NAME=my-shoot
    PROJECT_NAME=my-project

    kubectl patch -n shoot--${PROJECT_NAME}--${SHOOT_NAME} etcdcopybackupstask ${SHOOT_NAME} --subresource status --type merge -p "{\"status\":{\"conditions\":[{\"type\":\"Succeeded\",\"status\":\"True\",\"reason\":\"manual copy successful\",\"message\":\"manual copy successful\",\"lastTransitionTime\":\"$(date -Iseconds)\",\"lastUpdateTime\":\"$(date -Iseconds)\"}]}}"
    ```

> 1. After the `main-etcd` becomes `Ready`, remove the finalizer on the source `extensions.gardener.cloud/v1alpha1.BackupEntry` (the resource has `source-` as a prefix in its name) in the `Destination Seed` so that it can be deleted successfully. This is necessary as the `Destination Seed` will not have network connectivity to perform the actual deletion of the source backup directory.

Reviewer: > This is necessary as the `Destination Seed` will not have network connectivity to perform the actual deletion of the source backup directory.

In the next step this cleanup ought to happen manually. Doesn't this lead to the proper deletion of the source `BackupEntry`, without removing the finalizer?

Author: I think that the deletion of the `BackupEntry` resource as a whole would fail because the client won't be able to connect to the storage provider to do the deletion (e.g. for the aws provider - https://github.com/gardener/gardener-extension-provider-aws/blob/cb2d32524766e416dd11644ceb1b631389bb254e/pkg/controller/backupentry/actuator.go#L40-L45). This should be the case even if the backups were manually removed. If we delete the finalizer of the `extensions.BackupEntry` then.

However, I will add one more clarification to ensure that the `source-etcd-backup` has been removed (automatically) before removing the finalizer.
